File: sdks/typescript/examples/feedback-loop.ts
/**
 * Feedback loop — teach Engram which results your model actually
 * picked. Every time your LLM picks a subset of search results and
 * discards the rest, call `feedback()` with the IDs. Engram uses
 * the signal to boost selected hot-tier matches, penalize rejected
 * ones, and write PREFERRED_OVER edges in its internal graph — all
 * cost-free to you because your LLM already made the judgment.
 *
 * Run with:
 *   ENGRAM_API_KEY=pr_live_... npx tsx examples/feedback-loop.ts
 */

import { EngramClient } from "../src/index.js";

async function main(): Promise<void> {
  const apiKey = process.env.ENGRAM_API_KEY;
  if (!apiKey) throw new Error("Set ENGRAM_API_KEY first.");

  const client = new EngramClient({ apiKey });

  // Seed a handful of memories so the search has something to rank.
  const ids: string[] = [];
  const seeds = [
    "Staging API deploys through GitHub Actions on every merge to main.",
    "Prod deploys require a manual approval from an on-call engineer.",
    "Hotfixes skip staging via the fast-track workflow (infra/hotfix.yml).",
    "All deploys run the smoke-test suite before promoting traffic.",
  ];
  for (const text of seeds) {
    const r = await client.store({ text, category: "runbook" });
    ids.push(r.id);
  }

  // Search. Imagine your LLM then decides which results actually
  // answer the user's question.
  const query = "how do I deploy a hotfix to production";
  const hits = await client.search({ query, topK: 4 });
  console.log(`Got ${hits.results.length} results for "${query}"`);

  // Suppose the LLM judged results 0 and 2 as useful, 1 as off-topic.
  const selected = [hits.results[0], hits.results[2]].filter(Boolean);
  const rejected = [hits.results[1]].filter(Boolean);

  const fb = await client.feedback({
    query,
    selectedIds: selected.map((r) => r!.id),
    rejectedIds: rejected.map((r) => r!.id),
  });

  console.log("Feedback ingested:", {
    success: fb.success,
    boosted: fb.boosted,
    penalized: fb.penalized,
    edges_added: fb.edges_added,
  });

  // Next time someone asks a similar question, the boosted memories
  // will surface higher in the hot tier. No model finetune required.
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sdks/python/examples/feedback_loop.py
"""Minimal reranking feedback loop.

Simulates what a real agent loop does: search, let a (stub) judgment
function decide which hits are useful, then call /v1/feedback so the
cloud can boost good ids and penalize bad ones on future queries.

In production the judgment step is your own model — a cheap LLM
rerank, a rule-based classifier, or just the final state of your
context buffer. The key idea is that *any* signal you already
compute is useful to the cloud, so feedback should be cheap to wire
up everywhere search is called.

    export ENGRAM_API_KEY=pr_live_...
    python examples/feedback_loop.py
"""

from __future__ import annotations

import sys
from typing import Iterable, Tuple

from engram import (
    EngramClient,
    EngramError,
    SearchResult,
)


def simulated_judgment(
    query: str,
    hits: Iterable[SearchResult],
) -> Tuple[list[str], list[str]]:
    """Pretend to be a reranker.

    Returns (selected_ids, rejected_ids). Real code would call a
    model here or inspect which ids the final answer cited.

    The heuristic: keep anything above 0.6 score AND containing any
    non-trivial word from the query. Drop the rest. This is just a
    stand-in so the example is runnable offline without a second
    dependency, not a serious reranker.
    """
    keywords = {w.lower() for w in query.split() if len(w) > 3}
    selected: list[str] = []
    rejected: list[str] = []
    for hit in hits:
        text_tokens = {w.lower().strip(".,:;?!") for w in hit.text.split()}
        overlaps = bool(keywords & text_tokens)
        if hit.score >= 0.6 and overlaps:
            selected.append(hit.id)
        else:
            rejected.append(hit.id)
    return selected, rejected


def main() -> int:
    client = EngramClient()
    try:
        # Seed a handful of memories so the example has something to
        # rank. In a real app these would already be in the cloud.
        seed = [
            ("Production postgres lives on port 5433 with pgvector 0.7", "infra"),
            ("Staging postgres is on 5432 for legacy tooling", "infra"),
            ("We chose SQS FIFO for the primary queue", "decisions"),
            ("Mountain biking is fun but expensive", "personal"),
        ]
        for text, category in seed:
            client.store(text, category=category, importance=0.6)

        query = "what port does production postgres use"
        hits = client.search(query, top_k=5)
        print("[search] {} hits for {!r}".format(len(hits.results), query))
        for h in hits.results:
            print("  {:.2f} [{}] {}".format(h.score, h.tier or "?", h.text[:80]))

        selected, rejected = simulated_judgment(query, hits.results)
        print("[judge]  kept {} / dropped {}".format(len(selected), len(rejected)))

        result = client.feedback(
            query=query,
            selected_ids=selected,
            rejected_ids=rejected,
        )
        print(
            "[feedback] success={} boosted={} penalized={} edges_added={}".format(
                result.success,
                result.boosted,
                result.penalized,
                result.edges_added,
            )
        )
    except EngramError as exc:
        print("Engram call failed: {}".format(exc), file=sys.stderr)
        return 1
    finally:
        client.close()

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
